File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int getsol(vector<int>nums,int n,vector<int>&dp)
    {
        if(n<0)
        {
            return 0;
        }
        if(dp[n]!=-1)
        return dp[n];
        int take=nums[n]+getsol(nums,n-2,dp);
        int nottake=getsol(nums,n-1,dp);
        dp[n]= max(take,nottake);
        return dp[n];
    }
    int rob(vector<int>& nums) {
        int n=nums.size();
        vector<int>dp(n+1,-1);
        return getsol(nums,n-1,dp);

    }
};

File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int ans(int[] nums,int n,int index,int[] dp)
    {
        if(index>=n)
        return 0;
        if(dp[index]!=-1)
        return dp[index];
        int pick=nums[index]+ans(nums,n,index+2,dp);
        int notpick=ans(nums,n,index+1,dp);

        return dp[index]=Math.max(pick,notpick);

    }
    public int rob(int[] nums) {
        int n=nums.length;
        int[] dp=new int[n+1];
        Arrays.fill(dp,-1);
        return ans(nums,n,0,dp);
    }
}